File: src/frontend/src/api/liveTrading.ts
/**
 * 实盘交易 API
 */
import request from './index'

export interface LiveInstanceInfo {
  id: string
  strategy_id: string
  strategy_name: string
  status: 'running' | 'stopped' | 'error'
  pid: number | null
  error: string | null
  params: Record<string, unknown>
  created_at: string
  started_at: string | null
  stopped_at: string | null
  log_dir: string | null
}

export interface LiveInstanceListResponse {
  total: number
  instances: LiveInstanceInfo[]
}

export interface LiveBatchResponse {
  success: number
  failed: number
  details: { id: string; strategy_id: string; result: string }[]
}

export interface LiveGatewayPresetInfo {
  description?: string | null
  id: string
  name: string
  params: Record<string, unknown>
  editable_fields: LiveGatewayPresetFieldInfo[]
}

export interface LiveGatewayPresetFieldInfo {
  key: string
  label: string
  input_type: 'string' | 'boolean'
  placeholder?: string | null
}

export interface LiveGatewayPresetListResponse {
  total: number
  presets: LiveGatewayPresetInfo[]
}

export interface GatewayHealthInfo {
  gateway_key: string
  state: string
  is_healthy: boolean
  market_connection: string
  trade_connection: string
  uptime_sec: number
  last_heartbeat: number | null
  heartbeat_age_sec: number | null
  last_tick_time: number | null
  last_order_time: number | null
  strategy_count: number
  symbol_count: number
  tick_count: number
  order_count: number
  recent_errors: { timestamp: number; source: string; message: string }[]
  ref_count: number
  instances: string[]
  exchange?: string
  asset_type?: string
  account_id?: string
  strategy_name?: string
}

export interface GatewayHealthListResponse {
  total: number
  gateways: GatewayHealthInfo[]
}

export interface GatewayConnectRequest {
  exchange_type: string
  credentials: Record<string, unknown>
}

export interface GatewayConnectResponse {
  gateway_key: string
  status: string
  message: string
}

export const liveTradingApi = {
  list(): Promise<LiveInstanceListResponse> {
    return request.get('/live-trading/')
  },

  listPresets(): Promise<LiveGatewayPresetListResponse> {
    return request.get('/live-trading/presets')
  },

  add(strategy_id: string, params?: Record<string, unknown>): Promise<LiveInstanceInfo> {
    return request.post('/live-trading/', { strategy_id, params })
  },

  remove(instanceId: string): Promise<{ message: string }> {
    return request.delete(`/live-trading/${instanceId}`)
  },

  get(instanceId: string): Promise<LiveInstanceInfo> {
    return request.get(`/live-trading/${instanceId}`)
  },

  start(instanceId: string): Promise<LiveInstanceInfo> {
    return request.post(`/live-trading/${instanceId}/start`)
  },

  stop(instanceId: string): Promise<LiveInstanceInfo> {
    return request.post(`/live-trading/${instanceId}/stop`)
  },

  startAll(): Promise<LiveBatchResponse> {
    return request.post('/live-trading/start-all')
  },

  stopAll(): Promise<LiveBatchResponse> {
    return request.post('/live-trading/stop-all')
  },

  listGatewayHealth(): Promise<GatewayHealthListResponse> {
    return request.get('/live-trading/gateways/health')
  },

  connectGateway(data: GatewayConnectRequest): Promise<GatewayConnectResponse> {
    return request.post('/live-trading/gateways/connect', data)
  },

  disconnectGateway(gatewayKey: string): Promise<GatewayConnectResponse> {
    return request.post('/live-trading/gateways/disconnect', null, { params: { gateway_key: gatewayKey } })
  },

  listConnectedGateways(): Promise<{ total: number; gateways: { gateway_key: string; exchange_type: string; account_id: string; has_runtime: boolean }[] }> {
    return request.get('/live-trading/gateways/connected')
  },

  queryGatewayAccount(gatewayKey: string): Promise<Record<string, unknown>> {
    return request.get(`/live-trading/gateways/${encodeURIComponent(gatewayKey)}/account`)
  },

  queryGatewayPositions(gatewayKey: string): Promise<{ total: number; positions: Record<string, unknown>[] }> {
    return request.get(`/live-trading/gateways/${encodeURIComponent(gatewayKey)}/positions`)
  },
}

File: src/backend/app/types/live_trading.py
"""
Type definitions for live trading module.

Provides TypedDict and type aliases to replace Any types in live trading services.
"""

from typing import TypedDict

from typing_extensions import NotRequired


class InstanceData(TypedDict, total=False):
    """Live trading instance data structure."""

    id: str
    strategy_id: str
    user_id: str
    status: str  # pending, running, stopped, error
    pid: int
    created_at: str
    updated_at: str
    params: dict[str, str | int | float | bool]
    gateway_type: str
    gateway_key: str
    log_dir: str
    error: str


class GatewayCredentials(TypedDict, total=False):
    """Gateway connection credentials."""

    # CTP specific
    userid: str
    password: str
    brokerid: str
    td_address: str
    md_address: str
    appid: str
    auth_code: str
    # IB specific
    host: str
    port: int
    client_id: int
    # MT5 specific
    login: int
    server: str
    # Common
    exchange_type: str
    asset_type: str
    provider: str


class GatewayData(TypedDict, total=False):
    """Gateway runtime data."""

    key: str
    type: str
    status: str  # connected, disconnected, error
    gateway: object  # GatewayRuntime instance
    account_info: dict[str, str | float]
    positions: list[dict[str, str | float]]
    error: str
    pid: int


class GatewayPreset(TypedDict):
    """Gateway preset configuration."""

    type: str
    name: str
    exchange_type: str
    asset_type: str
    provider: str
    fields: list[dict[str, str]]


class StrategyConfig(TypedDict, total=False):
    """Strategy configuration from config.yaml."""

    strategy: dict[str, str | int | float | bool]
    gateway: dict[str, str | int | float | bool]
    data: dict[str, str | int | float | bool]


class SubprocessEnv(TypedDict):
    """Environment variables for subprocess."""

    PYTHONPATH: str
    STRATEGY_ID: str
    INSTANCE_ID: str
    LOG_DIR: str
    GATEWAY_TYPE: NotRequired[str]
    GATEWAY_KEY: NotRequired[str]


class GatewayLaunchParams(TypedDict, total=False):
    """Parameters for launching a gateway."""

    gateway_type: str
    transport: str
    td_address: str
    md_address: str
    userid: str
    password: str
    brokerid: str
    appid: str
    auth_code: str
    host: str
    port: int
    client_id: int
    login: int
    server: str


class HealthStatus(TypedDict, total=False):
    """Gateway health status.

    Mirrors the dict produced by ``gateway.health`` snapshots and
    ``LiveTradingManager._build_subprocess_gateway_health``.
    """

    gateway_key: str
    state: str
    is_healthy: bool
    exchange: str
    asset_type: str
    account_id: str
    market_connection: str
    trade_connection: str
    uptime_sec: float
    last_heartbeat: float | None
    heartbeat_age_sec: float | None
    last_tick_time: float | None
    last_order_time: float | None
    strategy_count: int
    symbol_count: int
    tick_count: int
    order_count: int
    ref_count: int
    instances: list[str]
    recent_errors: list[dict[str, str]]


class InstanceListResult(TypedDict):
    """Result of listing instances."""

    instances: list[InstanceData]
    total: int


class OperationResult(TypedDict, total=False):
    """Result of an operation.

    Mirrors the dict shape returned by gateway/instance service helpers,
    which use ``status``/``message`` (and an optional ``gateway_key``).
    """

    status: str
    message: str
    gateway_key: str
    success: bool
    data: dict


class StartResult(OperationResult, total=False):
    """Result of starting an instance."""

    pid: int


class StopResult(OperationResult, total=False):
    """Result of stopping an instance."""

    was_running: bool


class ConnectResult(OperationResult, total=False):
    """Result of connecting a gateway."""

    key: str
